src/main/java/frc/robot/subsystems/VisionControl.java

```diff
@@ -1,70 +1,241 @@
     private double ballr = visionData.br;
     private boolean wait = visionData.waitForOtherRobot;
-    // thresholds
-    private double chassisThreshold = 10;
-    private double shooterThreshold = 10;
-
     // shooter variables
     private double desiredAngle;// angle of the shooter in degrees
     private double desiredPower;// desired shooter power in RPMS
     // chassis variables
-    private double forward_speed = 0; // speed front to back between 0-1
-    private double sidespeed = 0; // straife speed between 0-1
-    private double rotation = 0; // rotation speed between 0-1
-    // private Chassis chassis;
+    private double hoop_forward_speed = 0; // speed front to back between 0-1
+    private double hoop_sidespeed = 0; // straife speed between 0-1
+    private double hoop_rotation = 0; // rotation speed between 0-1
+    private double ball_forward_speed = 0; // speed front to back between 0-1
+    private double ball_sidespeed = 0; // straife speed between 0-1
+    private double ball_rotation = 0; // rotation speed between 0-1
+    private Chassis chassis;
+    //other variables
+    public boolean usingAuto = false;
+    private int invalid_ball_counter = 0;    
+    private final int invalid_ball_counter_threshold = 20;
+    private MachineLearning ml;
+    private path1 p1 = new path1();
+    private double skip[] = {0};
+    public double counter = 0;
+    private double kP = 0;
+    private boolean teachTheAI = true;
+    private double counterSpeed;
+    private double frontRightSpeed[] = {};
+    private double frontLeftSpeed[] = {};
+    private double backRightSpeed[] = {};
+    private double backLeftSpeed[] = {};
+
+    //we need to determine what to set these to
+    
+    // thresholds
+    private final double ballChassisThreshold = 2; //angle in degrees
+    private final double hoopChassisThreshold = 2; //angle in degrees
+    private final double shooterThreshold = 2;
+
+    private final double[] hoopRange = {3, 50};//min, max in feet
+
+    private final boolean SQUARE_DRIVER_INPUTS = true;
+    private final double kpasta = 0.3;
+    private final double ki = 0;
+    private final double kd = 0.001;
+
+
+    //edit these
+    private final boolean RECORD_PATH = true;
+    private final String FILE_NAME = "path1";
+    private String selector; 
+
+    
     // private Shooter shooter;
 
-    public VisionControl(Vision vision, VisionData visionData, OperatorInterface oi) {// , Chassis chassis, Shooter shooter}) {
+    public VisionControl(Vision vision, VisionData visionData, OperatorInterface oi, Chassis chassis, IMU imu) {// , Shooter shooter}) {
+        this.selector = "";
         this.oi = oi;
         this.vision = vision;
         this.visionData = visionData;
-        // this.chassis = chassis;
+        this.chassis = chassis;
+        this.imu = imu;
+        ml = new MachineLearning(RECORD_PATH, FILE_NAME);
+        hoopDistancePid = new PIDController(kpasta, ki, kd);
```

Comment:
I think we decided that hoop will also use only rotation to aim (not drive forward or backward). Range to target will determine the shooter wheel speed.